samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c

```diff
@@ -0,0 +1,824 @@
+{
+	int err;
+
+	err = mqtt_disconnect(&client);
+	if (err) {
+		LOG_ERR("ERROR: mqtt_disconnect %d", err);
+	}
+	slm_at_mqtt_uninit();
+
+	return err;
+}
+
+static int do_mqtt_publish(u16_t qos, u16_t retain,
+				u8_t *topic, size_t topic_len,
+				u8_t *msg, size_t msg_len)
+{
+	int err = -EINVAL;
+	struct mqtt_publish_param param;
+
+	if (qos <= MQTT_QOS_2_EXACTLY_ONCE) {
+		param.message.topic.qos = (u8_t)qos;
+	} else {
+		return err;
+	}
+
+	if (retain <= 1) {
+		param.retain_flag = (u8_t)retain;
+	} else {
+		return err;
+	}
+	param.message.topic.topic.utf8 = topic;
+	param.message.topic.topic.size = topic_len;
+	param.message.payload.data = msg;
+	param.message.payload.len = msg_len;
+	param.message_id = sys_rand32_get();
+	param.dup_flag = 0;
+
+	return mqtt_publish(&client, &param);
+}
+
+static int do_mqtt_subscribe(u16_t op,
+				u8_t *topic_buf,
+				size_t topic_len,
+				u16_t qos)
+{
+	int err = -EINVAL;
+	struct mqtt_topic subscribe_topic;
+
+	const struct mqtt_subscription_list subscription_list = {
+		.list = &subscribe_topic,
+		.list_count = 1,
+		.message_id = sys_rand32_get()
+	};
+
+	if (qos <= MQTT_QOS_2_EXACTLY_ONCE) {
+		subscribe_topic.qos = (u8_t)qos;
+	} else {
+		return err;
+	}
+	subscribe_topic.topic.utf8 = topic_buf;
+	subscribe_topic.topic.size = topic_len;
+
+	if (op == 1) {
+		err = mqtt_subscribe(&client, &subscription_list);
+	} else if (op == 0) {
+		err = mqtt_unsubscribe(&client, &subscription_list);
+	}
+
+	return err;
+}
+
+/**@brief handle AT#XMQTTCON commands
+ *  AT#XMQTTCON=<op>[,<url>,<port>[,<sec_tag>]]
+ *  AT#XMQTTCON?
+ *  AT#XMQTTCON=?
+ */
+static int handle_at_mqtt_connect(enum at_cmd_type cmd_type)
+{
+	int err = -EINVAL;
+
+	u16_t op;
+	size_t url_sz = MQTT_MAX_URL_LEN;
+	size_t cid_sz = MQTT_MAX_CID_LEN;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		if (at_params_valid_count_get(&at_param_list) == 0) {
+			return -EINVAL;
+		}
+		err = at_params_short_get(&at_param_list, 1, &op);
+		if (err < 0) {
+			return err;
+		}
+		if (op == 1) {
+			if (at_params_valid_count_get(&at_param_list) <= 4) {
+				return -EINVAL;
+			}
+			if (ctx.connected) {
+				return -EINPROGRESS;
+			}
+
+			memset(&ctx, 0, sizeof(ctx));
+
+			err = at_params_string_get(&at_param_list, 2,
+							ctx.cid, &cid_sz);
+			if (err < 0) {
+				return err;
+			}
+			ctx.cid[cid_sz] = '\0';
+			err = at_params_string_get(&at_param_list, 3,
+							ctx.url, &url_sz);
+			if (err < 0) {
+				return err;
+			}
+			ctx.url[url_sz] = '\0';
+			err = at_params_int_get(&at_param_list, 4, &ctx.port);
+			if (err < 0) {
+				return err;
+			}
+			if (at_params_valid_count_get(&at_param_list) == 6) {
+				err = at_params_int_get(&at_param_list, 5,
+							&ctx.sec_tag);
+				if (err < 0) {
+					return err;
+				}
+				ctx.sec_transport = true;
+			}
+			err = do_mqtt_connect();
+			break;
+		} else if (op == 0) {
+			LOG_DBG("Disconnect from broker");
+			err = do_mqtt_disconnect();
```

Comment:
There is a need to check ctx.connected in case of not connected yet

Reply:
Added.